bouncer/shared/utils/substrate.ts

```diff
@@ -200,7 +255,10 @@ const subscribeHeads = getCachedDisposable(
       const historicApi = await api.at(header.hash);
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       const rawEvents = (await historicApi.query.system.events()) as unknown as any[];
-      subject.next(mapEvents(rawEvents, header.number.toNumber()));
+      const mappedEvents = mapEvents(rawEvents, header.number.toNumber());
+      const cache = eventCacheMap[chain];
+      cache.addEvents(header.number.toNumber(), mappedEvents);
```

Comment:
Might be more correct to cache by block hash rather than number? Shouldn't matter for localnet, but in case this is used on a real network where reorgs are possible, this might be problematic.

Reply:
This was my original plan but it turned out to be harder and slower. I would have to fetch each block hash every time we do a block number lookup to check hashes match. We would also have to maintain an extra list that maps block numbers to hash's so we can delete the oldest blocks still. So i thought it would be fine to just do block numbers for version 1 and avoid the complexity and overhead. Happy to upgrade it later.